File: VB/LocalizableDisplayName/LocalizableDisplayNameAttribute.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Globalization
Imports System.Reflection
Imports System.Resources

Public Class LocalizableDisplayNameAttribute
    Inherits DisplayNameAttribute
    Public Shared GlobalResourceCulture As CultureInfo = Nothing
    Private Shared ReadOnly ResourceManagers As New Dictionary(Of Type, ResourceManager)()
    Private ReadOnly resXGeneratedClass As Type
    Public Sub New(ByVal resXGeneratedClass As Type, ByVal stringName As String)
        MyBase.New(stringName)
        If Not ResourceManagers.ContainsKey(resXGeneratedClass) Then
            Dim resourceManager As ResourceManager = CType(resXGeneratedClass.InvokeMember("ResourceManager", BindingFlags.Static Or BindingFlags.NonPublic Or BindingFlags.GetProperty, Nothing, Nothing, Nothing), ResourceManager)
            ResourceManagers.Add(resXGeneratedClass, resourceManager)
        End If
        Me.resXGeneratedClass = resXGeneratedClass
    End Sub
    Public Overrides ReadOnly Property DisplayName As String
        Get
            Dim resourceManager As ResourceManager = ResourceManagers(resXGeneratedClass)
            Return resourceManager.GetString(MyBase.DisplayName, GlobalResourceCulture)
        End Get
    End Property
End Class

File: CS/LocalizableDisplayName/LocalizableDisplayNameAttribute.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.Reflection;
using System.Resources;

namespace LocalizableDisplayName {
    public class LocalizableDisplayNameAttribute : DisplayNameAttribute {
        public static CultureInfo GlobalResourceCulture = null;
        static readonly IDictionary<Type, ResourceManager> ResourceManagers = new Dictionary<Type, ResourceManager>();
        readonly Type resXGeneratedClass;
        public LocalizableDisplayNameAttribute(Type resXGeneratedClass, string stringName) : base(stringName) {
            if(!ResourceManagers.ContainsKey(resXGeneratedClass)) {
                ResourceManager resourceManager = (ResourceManager)resXGeneratedClass.InvokeMember("ResourceManager", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.GetProperty, null, null, null);
                ResourceManagers.Add(resXGeneratedClass, resourceManager);
            }
            this.resXGeneratedClass = resXGeneratedClass;
        }
        public override string DisplayName {
            get {
                ResourceManager resourceManager = ResourceManagers[resXGeneratedClass];
                return resourceManager.GetString(base.DisplayName, GlobalResourceCulture);
            }
        }
    }
}